### DISCUSSION TITLE
Statistic page is affected by MUI v5 or permissions plugin?

### DISCUSSION THREAD

**Discussion by ahmm16** [drodil/backstage-plugin-qeta#242] [General]
Hi all! I'm having problems with the latest versions of the plugin. This problem is related with styles and some pages like Statistics, profile,

When I want to go to these pages I receive this error: 

TypeError
stats.summary is undefined.
![statistic-errors](https://github.com/user-attachments/assets/f15fcdb2-aa59-43e0-ad34-4acd755e7302)

Regarding with styles, I dont know why I dont see the same or almost the same styles like in the documentation (left menu).

![left-menu-styles](https://github.com/user-attachments/assets/bdf2729f-45f2-4eb7-b23d-302b9920b571)

**Comment by drodil** [comment]
Hi,

Do you run the same version of the backend and the frontend? The stats issue was in some older backend version so just want to know if it has reproduced.

Styling issue is different and it's related to MUI5 update. Are you using custom theme? It looks totally different from the dev environment, all the tiles and icons look misplaced.

**Reply by drodil** [reply]
@ahmm16 I installed the plugin to new backstage app and it looks fine:

<img width="1697" alt="image" src="https://github.com/user-attachments/assets/dfa301b9-1a6a-4e49-8be4-bcee80278395">

So there must be something wrong in your app styling that doesn't support the MUI5 components. Any idea what could it be?

**Reply by ahmm16** [reply]
Hi @drodil thank you for your answers. 

Just give a couple of minutes to send you the versions. Anyways, what plugins are you using and with which versions?

**Reply by ahmm16** [reply]
These are my versions: 

Frontend package.json: 
 "@drodil/backstage-plugin-qeta": "^3.12.0",
 "@drodil/backstage-plugin-qeta-react": "^3.12.0",
 
Backend package.json:
"@drodil/backstage-plugin-qeta-backend": "^3.12.0",
"@drodil/backstage-plugin-qeta-common": "^3.12.0",
 
 
 Regarding with styles: Yes, I created a UnifiedTheme and Im just changing the palette properties with primary, background, banner and buttons colors, no more. I did a test which consisted of removing the entire UnifiedTheme but I still got the same errors.

**Reply by ahmm16** [reply]
Regarding this comment: _"So there must be something wrong with your app's style that doesn't support MUI5 components. Do you have any idea what it could be?"_

I haven't followed this documentation yet: https://backstage.io/docs/tutorials/migrate-to-mui5/ Do you think it might affect it?

**Reply by drodil** [reply]
I tried this on our own app with custom theme and the styles do not work. We have migrated the app to MUI5 so have to dig deeper what is going on as it works fine with plain backstage. If you come up with solution, please let me know too 😄

In the meanwhile, I will check if I can reproduce the statistics issue!

**Reply by drodil** [reply]
And the versions are 3.12.0 too

**Reply by ahmm16** [reply]
_I tried this on our own app with custom theme and the styles do not work. We have migrated the app to MUI5 so have to dig deeper what is going on as it works fine with plain backstage. If you come up with solution, please let me know too 😄

In the meanwhile, I will check if I can reproduce the statistics issue!_

Perfect @drodil . Thank you a lot. If I found solutions for that I will share with you for sure!

**Reply by drodil** [reply]
I just tried some fixes, new version 1.32.1 is soon out, please test if it helps at all!

**Reply by ahmm16** [reply]
Sorry @drodil but 1.32.1? Is it a hotfix in an older version? Or do you mean 3.12.1? 


Thank you a lot again! I will test it!

**Reply by drodil** [reply]
Yes sorry, mis spelled, 3.12.1 😄

**Reply by ahmm16** [reply]
Hi again @drodil . I tested with these versions but I didnt see any change: 

packages/app/package.json: 

"@backstage/theme": "^0.5.6",
"@drodil/backstage-plugin-qeta": "^3.12.0",
"@drodil/backstage-plugin-qeta-react": "^3.12.1",
"@material-ui/core": "^4.12.4",
"@material-ui/icons": "^4.11.3",
"react": "^17.0.2",
"react-dom": "^17.0.2",
"react-router": "^6.3.0",
"react-router-dom": "^6.3.0"

profile and statistics pages are giving me these errors: 

http://localhost:3000/qeta/users/user:default/jane:
TypeError
stats.summary is undefined
Call Stack
 SummaryStatsGrid
  node_modules/@drodil/backstage-plugin-qeta/node_modules/@drodil/backstage-plugin-qeta-react/dist/components/SummaryStatsGrid/SummaryStatsGrid.esm.js:38:7
 renderWithHooks
  node_modules/react-dom/cjs/react-dom.development.js:14985:27
 mountIndeterminateComponent
  node_modules/react-dom/cjs/react-dom.development.js:17811:13
  
http://localhost:3000/qeta/statistics:
TypeError
stats.summary is undefined
Call Stack
 SummaryStatsGrid
  node_modules/@drodil/backstage-plugin-qeta/node_modules/@drodil/backstage-plugin-qeta-react/dist/components/SummaryStatsGrid/SummaryStatsGrid.esm.js:38:7
 renderWithHooks
  node_modules/react-dom/cjs/react-dom.development.js:14985:27
 mountIndeterminateComponent
  node_modules/react-dom/cjs/react-dom.development.js:17811:13
 beginWork
  node_modules/react-dom/cjs/react-dom.development.js:19049:16

**Reply by drodil** [reply]
Can you check the response from statistics request?

**Reply by drodil** [reply]
Also update the qeta plugin to 3.12.1 and see if it helps

**Reply by ahmm16** [reply]
Sure, give me a couple of minutes, Im updating more dependencies too!

**Reply by drodil** [reply]
Btw, you don't need to add the react plugin as dependency unless you are using some components in for example your entity pages.

**Reply by ahmm16** [reply]
Yes, I thought that but I want to enable Q&A in entities page. Im going to do some test if it doens't work I will try to fix versions that works with new permissions (I need customize the permissions for some users) and the old look and fell.

**Reply by drodil** [reply]
Stats should not relate to permissions. Did you check the response from the summary request? I cannot reproduce it even with empty database.

**Reply by drodil** [reply]
If it's not possible to get the MUI5 work with custom theme, it might be that I have to rollback to MUI4. I don't know what is going on in the Backstage internals and why it doesn't work out of the box...

**Reply by ahmm16** [reply]
I saw you delivered a new version 3.12.2 and Im going to test it. Also I will give you an answer about summary request!

**Reply by ahmm16** [reply]
Opening: http://localhost:3000/qeta/statistics
Results: 
http://localhost:7007/api/qeta/statistics/global 500 Internal Server Error

{
    "error": {
        "name": "TypeError",
        "message": "g.date.toDateString is not a function",
        "stack": "TypeError: g.date.toDateString is not a function\n    at <anonymous> (/workspace/.../node_modules/@drodil/backstage-plugin-qeta-backend/src/service/routes/statistics.ts:66:18)\n    at Array.map (<anonymous>)\n    at <anonymous> (/workspace/.../node_modules/@drodil/backstage-plugin-qeta-backend/src/service/routes/statistics.ts:65:36)"
    },
    "request": {
        "method": "GET",
        "url": "/statistics/global"
    },
    "response": {
        "statusCode": 500
    }
}

These 3 dots here: /workspace/.../node_modules/ is just to avoid share organizational data (my boss can kill me xD hahaha)

**Reply by drodil** [reply]
Ok, are you running this on postgres or sqlite?

**Reply by ahmm16** [reply]
sqlite

Also, trying to update libraries and Im receiving this:

Module not found: Can't resolve '@mui/icons-material/Flare' in '/workspace/eric-csdp-portal/node_modules/@drodil/backstage-plugin-qeta/node_modules/@drodil/backstage-plugin-qeta-react/dist/components/AIAnswerCard'
node_modules/@drodil/backstage-plugin-qeta/node_modules/@drodil/backstage-plugin-qeta-react/dist/components/AIAnswerCard/AIAnswerCard.esm.js

**Reply by drodil** [reply]
Ah, must be because of knex is not returning date object for the field for sqlite. No idea why it cannot find the flare icon, it's supported in MUI (https://fonts.google.com/icons?icon.set=Material+Icons&icon.query=flare)

**Reply by drodil** [reply]
Flare icon is also supported in MUI4 so it should be fine, can you check your node_modules if the flare is available there?

**Comment by ahmm16** [comment]
Lets remove node_modules and yarn install again to see it can be solved. Regarding with sqlite things is strange because previous versions worked fine =/

**Reply by drodil** [reply]
I bet it's because you now have some statistics in the database and previously didn't?

**Reply by drodil** [reply]
I forced the date to be Date object now and will release yet another version, hope it fixes this. I don't really run this plugin with sqlite anywhere so these things don't get caught so easily :D

**Reply by ahmm16** [reply]
_I bet it's because you now have some statistics in the database and previously didn't?_

- No no, in fact, I'm using older versions of Q&A in production and its working properly with some data.

**Reply by drodil** [reply]
I mean that if you now locally have some data in the database, it will try to map those here https://github.com/drodil/backstage-plugin-qeta/blob/main/plugins/qeta-backend/src/service/routes/statistics.ts#L67 and because that was probably a string, it could not call the `toDateString`. But hope it works now as I changed it to be Date always.

**Reply by ahmm16** [reply]
Good news man! Statistics and profile pages are working properly!

**Reply by ahmm16** [reply]
Now I only have problems related with styles of the left menu:

![left-menu-problems](https://github.com/user-attachments/assets/daa74877-6b85-45b5-bda4-4b141a72ab7f)

**Reply by drodil** [reply]
Yeah, I noticed the same issue and had to override the `MuiButtonBase` in the custom theme. But that doesn't fix everything; for example the Grid items are without padding etc. 

It's starting to feel more and more that I should rollback the MUI5 update... Though it's a lot of work to revert everything..

**Reply by ahmm16** [reply]
I can imagine that, I would actually love to help you but I am still learning a lot of things about Backstage and in our project we still have many open fronts =/. Let me know if you do the downgrade it will be very helpful for us!

And thank you so so much!

**Reply by drodil** [reply]
@ahmm16 I will check this a bit more next week and if no solution is found, I think rollback is the only option to be able to continue feature development. 

I probably try what happens if I add minimal custom theme to plain backstage instance and create an issue to the project if it really breaks. Have to read also some backstage code to understand more how themes work internally in the system. It does some internal magic to support both MUI4 and 5..

**Comment by drodil** [comment]
See also #244

**Comment by drodil** [comment]
Decided to roll back to MUI4. Details in #244

**Reply by ahmm16** [reply]
![fix-error](https://github.com/user-attachments/assets/58294abc-f83e-4d34-93e5-47a8cf03f8b8)

What a great change @drodil . Thank you so so much!!!
